// option-t/lib/Maybe/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Maybe_1 = require("./Maybe");
exports.isNotNullAndUndefined = Maybe_1.isNotNullAndUndefined;
exports.isNullOrUndefined = Maybe_1.isNullOrUndefined;
// XXX: `and()` operation is equivalent of `a && b` so we don't ship it by default set.
//export { andForMaybe as and } from './and';
var andThen_1 = require("./andThen");
exports.andThen = andThen_1.andThenForMaybe;
var expect_1 = require("./expect");
exports.expect = expect_1.expectNotNullAndUndefined;
var map_1 = require("./map");
exports.map = map_1.mapForMaybe;
var mapOr_1 = require("./mapOr");
exports.mapOr = mapOr_1.mapOrForMaybe;
var mapOrElse_1 = require("./mapOrElse");
exports.mapOrElse = mapOrElse_1.mapOrElseForMaybe;
// XXX: `or()` operation is equivalent of `a || b` so we don't ship it by default set.
// export { orForMaybe as or } from './or';
var orElse_1 = require("./orElse");
exports.orElse = orElse_1.orElseForMaybe;
var tap_1 = require("./tap");
exports.tap = tap_1.tapMaybe;
var unwrap_1 = require("./unwrap");
exports.unwrap = unwrap_1.unwrapMaybe;
var unwrapOr_1 = require("./unwrapOr");
exports.unwrapOr = unwrapOr_1.unwrapOrFromMaybe;
var unwrapOrElse_1 = require("./unwrapOrElse");
exports.unwrapOrElse = unwrapOrElse_1.unwrapOrElseFromMaybe;
