// stylelint-no-unused-selectors/dist/utils/document-resolver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const util_1 = require("util");
const string_template_1 = __importDefault(require("string-template"));
const readFile = util_1.promisify(fs_1.default.readFile);
function resolveDocuments(cssPath, documents) {
    const parsed = path_1.default.parse(cssPath);
    const resolved = documents.map((doc) => string_template_1.default(doc, {
        cssDir: parsed.dir,
        cssDirName: path_1.default.basename(parsed.dir),
        cssName: parsed.name,
    }));
    return resolved;
}
exports.resolveDocuments = resolveDocuments;
async function resolveDocument(paths) {
    for (const path of paths) {
        try {
            const content = await readFile(path, { encoding: 'utf-8' });
            return {
                path,
                document: content.toString(),
            };
        }
        catch (_a) { } // eslint-disable-line no-empty
    }
    return undefined;
}
exports.resolveDocument = resolveDocument;
