// option-t/cjs/Nullable/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Nullable_1 = require("./Nullable");
exports.isNotNull = Nullable_1.isNotNull;
exports.isNull = Nullable_1.isNull;
// XXX: `and()` operation is equivalent of `a && b` so we don't ship it by default set.
//export { andForNullable as and } from './and';
var andThen_1 = require("./andThen");
exports.andThen = andThen_1.andThenForNullable;
var expect_1 = require("./expect");
exports.expect = expect_1.expectNotNull;
var map_1 = require("./map");
exports.map = map_1.mapForNullable;
var mapOr_1 = require("./mapOr");
exports.mapOr = mapOr_1.mapOrForNullable;
var mapOrElse_1 = require("./mapOrElse");
exports.mapOrElse = mapOrElse_1.mapOrElseForNullable;
// XXX: `or()` operation is equivalent of `a || b` so we don't ship it by default set.
// export { orForNullable as or } from './or';
var orElse_1 = require("./orElse");
exports.orElse = orElse_1.orElseForNullable;
var tap_1 = require("./tap");
exports.tap = tap_1.tapNullable;
var unwrap_1 = require("./unwrap");
exports.unwrap = unwrap_1.unwrapNullable;
var unwrapOr_1 = require("./unwrapOr");
exports.unwrapOr = unwrapOr_1.unwrapOrFromNullable;
var unwrapOrElse_1 = require("./unwrapOrElse");
exports.unwrapOrElse = unwrapOrElse_1.unwrapOrElseFromNullable;
