// option-t/cjs/Undefinable/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Undefinable_1 = require("./Undefinable");
exports.isNotUndefined = Undefinable_1.isNotUndefined;
exports.isUndefined = Undefinable_1.isUndefined;
// XXX: `and()` operation is equivalent of `a && b` so we don't ship it by default set.
//export { andForUndefinable as and} from './and';
var andThen_1 = require("./andThen");
exports.andThen = andThen_1.andThenForUndefinable;
var expect_1 = require("./expect");
exports.expect = expect_1.expectNotUndefined;
var map_1 = require("./map");
exports.map = map_1.mapForUndefinable;
var mapOr_1 = require("./mapOr");
exports.mapOr = mapOr_1.mapOrForUndefinable;
var mapOrElse_1 = require("./mapOrElse");
exports.mapOrElse = mapOrElse_1.mapOrElseForUndefinable;
// XXX: `or()` operation is equivalent of `a || b` so we don't ship it by default set.
// export { orForUndefinable as or} from './or';
var orElse_1 = require("./orElse");
exports.orElse = orElse_1.orElseForUndefinable;
var tap_1 = require("./tap");
exports.tap = tap_1.tapUndefinable;
var unwrap_1 = require("./unwrap");
exports.unwrap = unwrap_1.unwrapUndefinable;
var unwrapOr_1 = require("./unwrapOr");
exports.unwrapOr = unwrapOr_1.unwrapOrFromUndefinable;
var unwrapOrElse_1 = require("./unwrapOrElse");
exports.unwrapOrElse = unwrapOrElse_1.unwrapOrElseFromUndefinable;
